### ISSUE TITLE
Ability to create SQL table logic from schema without doing a "migrate"

### ISSUE THREAD

**Issue by cmidgley** [prisma/prisma#6155]
## Problem

We use the SQL file generated by Prisma Migrate directly, without having Prisma migrate the database itself.  We do this so we can apply the SQL table generation logic in our test suites to generate ten or hundreds of schemas rapidly, on-demand.  Calling out to the Prisma executable is very slow (3-5 seconds per schema) whereas by directly applying the SQL in a Query (Postgres in our case) we can create our schemas in a fraction of a second.  Currently we have a build process that first destroys the files under Migrate, then asks Prisma Migrate (with ---create-only --skip-generate --name generate) to do the "migration", and then we then need to scan the migrations folder to find the generated SQL file.  This is a bit hacky, and more importantly, slow.  With this change, we reduced our test time from 4-6 seconds per test down to 0.5-1 second.

## Suggested solution

Adjust the new Prisma DB PUSH command to offer a feature that does not talk to the database and let's us specify the target location of the full SQL schema generation file.  For example:

```
prisma db push --preview-feature --sql-only --sql-name ./sql/generate-schema.sql
```

**Comment by janpio** [comment]
> Calling out to the Prisma executable is very slow (3-5 seconds per schema)

That sounds unwanted and unexpected. Can you please also create an issue for this so we can look into it? Thanks.

**Comment by cmidgley** [comment]
> That sounds unwanted and unexpected. Can you please also create an issue for this so we can look into it? 

I'm not sure I understand why 3-5 seconds is unexpected for the Prisma CLI to do a migration, and I'd rather not create an issue on it unless for some reason others find it is radically faster.

We call it using ([execSync](https://nodejs.org/api/child_process.html#child_process_child_process_execsync_command_options)) - perhaps there is an API that we can consume that Prisma CLI uses to generate the SQL instead?  That would be even faster/better.

Thanks!

**Comment by janpio** [comment]
> I'm not sure I understand why 3-5 seconds is unexpected for the Prisma CLI to do a migration, and I'd rather not create an issue on it unless for some reason others find it is radically faster.

I thought you gave the reason yourself - running the generated SQL file manually is much faster than calling our CLI.

**Comment by cmidgley** [comment]
I'm sorry I'm not being clear.  Right now, as best I am aware, the CLI only generates a SQL file with full table definitions, when it is asked to do a migration  - where there are no prior migrations and creating the SQL file is an unknown subdirectory under the 'migrations' folder.

I am asking for a feature where I can ask the CLI to generate a SQL file with full table definitions on-demand and specify where that file should be placed, where it does not make be destroy my prior migration entries to do so (and would also not need to talk to my database to examine it for matching schema).

My point about it being slow isn't that the CLI is slow, it's that it is slow to use it tens or hundreds of time on every test run.  By asking it once to create the SQL file, that file can be easily played into the target database efficiently in the test harness.  

So the ask isn't to improve performance, it is to add an option to `db push` (or `migrate`) to create a SQL rather directly, rather than pushing it into a database, or making me remove the existing migrations to get it.

Hopefully this makes more sense.  Sorry for the confusion.

**Comment by janpio** [comment]
Ah, I did indeed not get that fully before - sorry. Thanks for clarifying.

If I understood correctly there are effectively two parts to your suggestion:

1. Be able instruct the command to assume you have an empty database (and not try to talk to the database to figure out if that is actually true)
2. Instead of trying to execute that file towards the database, just write it to a file to be used in any way you want it to.

The second point can already kinda be achieved with `prisma migrate dev --create-only` - but it will always try to figure out the current state of the database first.

**Comment by cmidgley** [comment]
Yes, that's it!   Plus perhaps:

3. Ignore the existing migration folder and files, and allow the SQL file to be written to any path

I do use `--create-only` today ... and assuming I delete the migration folders, it all works.  Felt like an obvious extension to the `db push` feature rather than `migrate`,  and one that didn't seem too difficult (easily said when I'm not the author)!

**Comment by do4gr** [comment]
Hey, I know it's been a couple of months since you asked for this @cmidgley, but we now have a solution that we believe will work for you with two new migrate commands released in 3.9 https://github.com/prisma/prisma/releases/tag/3.9.0.

If I understood you correctly you want the full SQL to get from an empty database to the state of your Prisma schema. 

` npx prisma migrate diff --preview-feature --from-empty --to-schema-datamodel ./schema.prisma --script > script.sql`
will give you this. 

The second command will apply this to a database.

` npx prisma db execute --preview-feature --url "$DATABASE_URL_PROD" --file script.sql` 

Please let us know when you had the opportunity to try this out.
